### DISCUSSION TITLE
Unable to Dump Game Genie ROM (Genesis/Mega Drive)

### DISCUSSION THREAD

**Discussion by PW5190** [sanni/cartreader#719] [Need help]
Hey, I have a problem with my SEGA Genesis Game Genie cheating device.

I put it into the SEGA Genesis/MD slot, set the volt switch to 5V, and left the EEP, CLK0, and CLK1 switches off.

I then turned on my OSCR, and navigated to the `Mega Drive` section, then selected `Game Cartridge`. Here's the stats:

```
Cart Info:

Name: GameGenie
bramCheck: 0100
Size: 0 MBit
ChkS: 0000
Sram: None
```

I then pressed `Read Rom`, and I got an error:
```
Saving to MD/ROM/Game
Genie/47/...
[*******************]
Internal checksum...E
rror
37C7

CRC: Glitched
```

I tried it both with and without another Genesis game on top, and it won't dump the ROM.

**Comment by sanni** [comment]
Try updating to [12.4](https://github.com/sanni/cartreader/archive/refs/heads/master.zip) to see if that changes anything.

**Reply by PW5190** [reply]
I'm having a little trouble manually updating the firmware for my OSCR. Is there a way to automate the update process?

**Reply by sanni** [reply]
Easiest way would be to download [V12.3_Portable.zip](https://github.com/sanni/cartreader/releases/download/V12.3/V12.3_Portable.zip) extract it, go into the V12.3_Portable\Arduino IDE Portable\portable\sketchbook\Cart_Reader\ directory and delete all files there. Then copy the same files from [12.4](https://github.com/sanni/cartreader/archive/refs/heads/master.zip) into that directory and follow the readme.txt

**Reply by PW5190** [reply]
OK, I updated, and was able to dump the Game Genie ROM. The OSCR still thinks there's an error, but it dumped anyway. I threw the ROM in Kega Fusion, and discovered the checksum for the ROM was incorrect, but it did boot into the code entry page.

**Comment by sanni** [comment]
In the no-intro set there are two Game Genie dumps, both are 32KB in size.

Game Genie (USA, Europe) (Rev A) (Program).md 
CRC32: 14DBCE4A

Game Genie (USA, Europe) (Program).md
CRC32: 5F293E4C

Your dump probably is 128KB since that is the smallest size I set. So what you can try is to to change both occurrences of 0x20000 to 0x8000 in [MD.ino line 1097](https://github.com/sanni/cartreader/blob/master/Cart_Reader/MD.ino#L1097)

Change:
````
  // Some games are missing the ROM size in the header, in this case calculate ROM size by looking for mirror of the first line of the ROM
  // This does not work for cartridges that have SRAM mapped directly after the maskrom like Striker (Europe)
  if ((cartSize < 0x8000) || (cartSize > 0xEAF400)) {
    for (cartSize = 0x20000 / 2; cartSize < 0x400000 / 2; cartSize += 0x20000 / 2) {
      if ((readWord_MD(0x0) == readWord_MD(cartSize)) && (readWord_MD(0x1) == readWord_MD(0x1 + cartSize)) && (readWord_MD(0x2) == readWord_MD(0x2 + cartSize)) && (readWord_MD(0x3) == readWord_MD(0x3 + cartSize)) && (readWord_MD(0x4) == readWord_MD(0x4 + cartSize)) && (readWord_MD(0x5) == readWord_MD(0x5 + cartSize)) && (readWord_MD(0x6) == readWord_MD(0x6 + cartSize)) && (readWord_MD(0x7) == readWord_MD(0x7 + cartSize)) && (readWord_MD(0x8) == readWord_MD(0x8 + cartSize))) {
        break;
      }
    }
    cartSize = cartSize * 2;
  }
````

to:
````
  // Some games are missing the ROM size in the header, in this case calculate ROM size by looking for mirror of the first line of the ROM
  // This does not work for cartridges that have SRAM mapped directly after the maskrom like Striker (Europe)
  if ((cartSize < 0x8000) || (cartSize > 0xEAF400)) {
    for (cartSize = 0x8000 / 2; cartSize < 0x400000 / 2; cartSize += 0x8000 / 2) {
      if ((readWord_MD(0x0) == readWord_MD(cartSize)) && (readWord_MD(0x1) == readWord_MD(0x1 + cartSize)) && (readWord_MD(0x2) == readWord_MD(0x2 + cartSize)) && (readWord_MD(0x3) == readWord_MD(0x3 + cartSize)) && (readWord_MD(0x4) == readWord_MD(0x4 + cartSize)) && (readWord_MD(0x5) == readWord_MD(0x5 + cartSize)) && (readWord_MD(0x6) == readWord_MD(0x6 + cartSize)) && (readWord_MD(0x7) == readWord_MD(0x7 + cartSize)) && (readWord_MD(0x8) == readWord_MD(0x8 + cartSize))) {
        break;
      }
    }
    cartSize = cartSize * 2;
  }
````

And then see if it dumps a 32KB ROM.
